src/shims/env.rs

```diff
@@ -150,4 +154,30 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
             }
         }
     }
+
+    fn update_environ(&mut self) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+        // Collect all the pointers to each variable in a vector.
+        let mut vars: Vec<Scalar<Tag>> = this.machine.env_vars.map.values().map(|&ptr| ptr.into()).collect();
+        // Add the trailing null pointer.
+        vars.push(Scalar::from_int(0, this.pointer_size()));
+        // Make an array with all these pointers inside Miri.
+        let tcx = this.tcx;
+        let environ_layout =
+            this.layout_of(tcx.mk_array(tcx.mk_imm_ptr(tcx.types.u8), vars.len() as u64))?;
+        let environ_place = this.allocate(environ_layout, MiriMemoryKind::Machine.into());
+        for (idx, var) in vars.into_iter().enumerate() {
+            let place = this.mplace_field(environ_place, idx as u64)?;
+            this.write_scalar(var, place.into())?;
+        }
+
+        let scalar: Scalar<Tag> = environ_place.ptr.into();
+
+        this.write_scalar(
+            scalar,
```

Comment:
`scalar` is such a nondescript variable name, I think you are better of just inlining its definition here.

Reply:
In fact, `write_scalar` already takes an `Into<ScalarMaybeUndef>`, so you can scratch the `.into()` as well.